### ISSUE TITLE
clippy::cargo lint is applied to sibling crates

### ISSUE THREAD

**Issue by purplesyringa** [rust-lang/rust-clippy#13745]
### Summary

When `#[warn(clippy::cargo)]` is enabled in one crate in a workspace, warnings are emitted for other crates in the same workspace too.

### Reproducer

I tried this code:

```rust
.
├── a
│  ├── Cargo.toml
│  └── src
│     └── main.rs
├── b
│  ├── Cargo.toml
│  └── src
│     └── main.rs
├── Cargo.lock
└── Cargo.toml
```

`a/src/main.rs`:

```rust
#![warn(clippy::cargo)]

fn main() {
    println!("Hello, world!");
}
```

`b/src/main.rs`:

```rust
fn main() {
    println!("Hello, world!");
}
```

I expected to see this happen: `cargo clippy` emits warnings for `a` but not `b`

Instead, this happened: `cargo clippy` emits warnings for both `a` and `b`

### Version

```text
rustc 1.84.0-nightly (1e4f10ba6 2024-10-29)
binary: rustc
commit-hash: 1e4f10ba6476e48a42a79b9f846a2d9366525b9e
commit-date: 2024-10-29
host: x86_64-unknown-linux-gnu
release: 1.84.0-nightly
LLVM version: 19.1.1
```

**Comment by blyxyas** [comment]
I've tested this, and seems that it doesn't lint on sibling crates anymore =^w^= do you have another example or do we close this as completed?

**Comment by purplesyringa** [comment]
It's still kind of broken, though the reproducer gets a bit uglier.

```
$ tree
.
├── a
│   ├── Cargo.toml
│   └── src
│       └── lib.rs
├── b
│   ├── Cargo.toml
│   └── src
│       └── lib.rs
├── Cargo.lock
└── Cargo.toml
```

Workspace-level `Cargo.toml` enables the lint:

```toml
[workspace]
members = ["a", "b"]
resolver = "2"

[workspace.lints.clippy]
cargo = "warn"
```

Crate A disables the lint in `a/src/lib.rs`:

```rust
#![allow(clippy::cargo)]
```

Crate B's `lib.rs` is empty.

`cargo clippy` warns about missing metadata fields for both A and B even though the lint is disabled for A.

---

More data: if the lints are disabled for both A and B, no warnings are shown. If neither A nor B disable the lints, lints are obviously emitted for both A and B, *but*: Clippy emits 12 warnings in total (so 6 for A and 6 for B) and then says

```
warning: `a` (lib) generated 12 warnings
warning: `b` (lib) generated 12 warnings (12 duplicates)
```

So what I think happening is that linting one crate emits warnings for *all* crates, and the "12 duplicates" here demonstrates this quadratic behavior.

**Comment by purplesyringa** [comment]
Looks like the whole `clippy::cargo` lint namespace is cursed. I'm assuming that [`LateLintPass::check_crate`](https://github.com/rust-lang/rust-clippy/blob/76118ec84ee5c1e865d10cd5d99c02878dd56ade/clippy_lints/src/cargo/mod.rs#L240) is invoked for each crate individually, but then [`common_metadata::check`](https://github.com/rust-lang/rust-clippy/blob/76118ec84ee5c1e865d10cd5d99c02878dd56ade/clippy_lints/src/cargo/common_metadata.rs#L9) iterates over each package, but the same applies to `feature_name`, too. `lint_groups_priority` reads `Cargo.toml` directly, and I don't know if that's correct, `multiple_crate_versions` works *correctly* but has superquadratic time complexity because it sorts `packages` on each iteration for some reason, and `wildcard_dependencies` only lints the first package.
